sway-fmt-v2/src/utils/pattern.rs

```diff
@@ -0,0 +1,116 @@
+use crate::fmt::*;
+use std::fmt::Write;
+use sway_parse::{token::Delimiter, Pattern, PatternStructField};
+use sway_types::Spanned;
+
+use super::bracket::{CurlyBrace, Parenthesis};
+
+impl Format for Pattern {
+    fn format(
+        &self,
+        formatted_code: &mut FormattedCode,
+        formatter: &mut Formatter,
+    ) -> Result<(), FormatterError> {
+        match self {
+            Self::Wildcard { underscore_token } => {
+                formatted_code.push_str(underscore_token.span().as_str())
+            }
+            Self::Var { mutable, name } => {
+                if let Some(mut_token) = mutable {
+                    write!(formatted_code, "{} ", mut_token.span().as_str())?;
+                }
+                // maybe add `Ident::format()`, not sure if needed yet.
+                formatted_code.push_str(name.span().as_str());
+            }
+            Self::Literal(lit) => lit.format(formatted_code, formatter)?,
+            Self::Constant(path) => path.format(formatted_code, formatter)?,
+            Self::Constructor { path, args } => {
+                path.format(formatted_code, formatter)?;
+                Self::open_parenthesis(formatted_code, formatter)?;
+                // need to add `<Pattern, CommaToken>` to `Punctuated::format()`
+                args.clone()
+                    .into_inner()
+                    .format(formatted_code, formatter)?;
+                Self::close_parenthesis(formatted_code, formatter)?;
+            }
+            Self::Struct { path, fields } => {
+                path.format(formatted_code, formatter)?;
+                Self::open_curly_brace(formatted_code, formatter)?;
+                // need to add `<PatternStructField, CommaToken>` to `Punctuated::format()`
+                fields
+                    .clone()
+                    .into_inner()
+                    .format(formatted_code, formatter)?;
+                Self::close_curly_brace(formatted_code, formatter)?;
+            }
+            Self::Tuple(args) => {
+                Self::open_parenthesis(formatted_code, formatter)?;
+                // need to add `<Pattern, CommaToken>` to `Punctuated::format()`
+                args.clone()
+                    .into_inner()
+                    .format(formatted_code, formatter)?;
+                Self::close_parenthesis(formatted_code, formatter)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+// Currently these just push their respective chars, we may need to change this
```

Comment:
W.r.t these `Parenthesis` and `CurlyBrace` traits, I agree it might be worth rethinking whether or not we want these as traits in general, particularly as we don't seem to be abstracting over their implementors in a generic way anywhere.

One way that we might be able to add some use to them could be to add a default method to the trait along the lines of:

```rust
trait Parenthesis {
    // open and close methods...

    /// Applies the given `format` function between a pair of open and close parentheses.
    fn parenthesised<F>(
        formatted_code: &mut FormattedCode,
        formatter: &mut Formatter,
        format: F,
    ) -> Result<(), FormatterError>
    where
        F: FnOnce(&mut FormattedCode, &mut Formatter) -> Result<(), FormatterError>,
    {
        self.open_parenthesis(formatted_code, formatter)?;
        format(formatted_code, formatter)?;
        self.close_parenthesis(formatted_code, formatter)
    }
}
```

This would allow us to do stuff like:

```rust
block.parenthesised(formatted_code, formatter, |formatted_code, formatter| {
    // format stuff inside block
})?;
```

where the open/close parens are automatically placed.

That said, it is still a bit awkward piping through the `formatted_code` and `formatter` and I'm not sure a fancy method like this is worth it?

Either way, these are considerations for a future PR - I think it's fine as is :)

Reply:
That's kinda cool actually, I wouldn't mind looking further into that.